presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LeftJoinNullFilterToSemiJoin.java

```diff
@@ -0,0 +1,150 @@
+import com.facebook.presto.spi.plan.Assignments;
+import com.facebook.presto.spi.plan.FilterNode;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.isRewriteLeftJoinNullFilterToSemiJoinEnabled;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.expressions.LogicalRowExpressions.and;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.sql.planner.VariablesExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.relational.Expressions.coalesceNullToFalse;
+import static com.facebook.presto.sql.relational.Expressions.constantNull;
+import static com.facebook.presto.sql.relational.Expressions.not;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.function.Function.identity;
+
+/**
+ * Rewrite Left Join with is null check on right join key to semi join, when other output from right input is not used.
+ * <pre>
+ *     - Filter
+ *          r.k is null
+ *          - Left Join
+ *              l.k = r.k
+ *              - Scan l
+ *                  k
+ *              - Scan r
+ *                  k
+ * </pre>
+ * to
+ * <pre>
+ *     - Project
+ *          l.k := l.k
+ *          r.k := NULL
+ *          - Filter
+ *              not(coalesce(semiJoinOutput, false))
+ *              - Semi Join
+ *                  source: l.k
+ *                  filter source: r.k
+ *                  filter output: semiJoinOutput
+ *                  - scan l
+ *                      k
+ *                  - scan r
+ *                      k
+ * </pre>
+ */
+public class LeftJoinNullFilterToSemiJoin
+        implements Rule<FilterNode>
+{
+    private static final Capture<JoinNode> CHILD = newCapture();
+    private static final Pattern<FilterNode> PATTERN = filter().with(source().matching(join().matching(x -> x.getType().equals(JoinNode.Type.LEFT)
+            && x.getCriteria().size() == 1 && !x.getFilter().isPresent() && x.getDynamicFilters().isEmpty()).capturedAs(CHILD)));
```

Comment:
How about if some fields from the rhs are used post-join? Do you check that somewhere? In that case, you cannot do this rewrite.